base/intfuncs.jl

```diff
@@ -195,29 +195,28 @@ end
 ^(x::Number, p::Integer)  = power_by_squaring(x,p)
 ^(x, p::Integer)          = power_by_squaring(x,p)
 
-# x^p for any literal integer p is lowered to x^Val{p},
+# x^p for any literal integer p is lowered to Base.literal_pow(^, x, Val{p})
 # to enable compile-time optimizations specialized to p.
-# However, we still need a fallback that calls the general ^.
-# To avoid ambiguities for methods that dispatch on the
-# first argument, we dispatch the fallback via internal_pow.
+# However, we still need a fallback that calls the function ^ which may either
+# mean Base.^ or something else, depending on context.
 # We mark these @inline since if the target is marked @inline,
 # we want to make sure that gets propagated,
 # even if it is over the inlining threshold.
-@inline ^(x, p) = internal_pow(x, p)
-@inline internal_pow{p}(x, ::Type{Val{p}}) = x^p
+@inline literal_pow{p}(^, x, ::Type{Val{p}}) = ^(x,p)
```

Comment:
I think I originally called it `y`, but Stefan suggested we change it to `^` earlier in this thread. I am fine with either but can change it to whatever the powers that be decide (pun intended)